### PR TITLE
Issue/47 null state handling

### LINKED ISSUE #47: Null state can make model unusable
There are two issues here:
1. If a resource deployment (create or update) fails, the state we receive from the provider might be null, in that case, we save it anyway, this is wrong.  A state can never be null.
2. If a state is null, and we use a plugin to get a value from that state, the compile will fail with the following error:
    ```
    ValueError: A string attribute can only get an item from a dict, but value is of type <class 'NoneType'>
    ```

<details>
<summary>Full error trace</summary>

```
Traceback (most recent call last):
  File "/var/lib/inmanta/server/environments/bde43af6-976a-4b57-a4b0-bc145a68f657/.env/lib64/python3.9/site-packages/inmanta_plugins/terraform/__init__.py", line 130, in get_resource_attribute
    return attribute_reference.extract_from_state(resource_state)
  File "/var/lib/inmanta/server/environments/bde43af6-976a-4b57-a4b0-bc145a68f657/.env/lib64/python3.9/site-packages/inmanta_plugins/terraform/helpers/attribute_reference.py", line 66, in extract_from_state
    raise ValueError(
ValueError: A string attribute can only get an item from a dict, but value is of type <class 'NoneType'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/ast/statements/call.py", line 241, in call_in_context
    value = self.plugin(*args, **kwargs)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/plugins.py", line 431, in __call__
    value = self.call(*new_args, **new_kwargs)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/plugins.py", line 512, in wrapper
    return fnc(*args, **kwargs)
  File "/var/lib/inmanta/server/environments/bde43af6-976a-4b57-a4b0-bc145a68f657/.env/lib64/python3.9/site-packages/inmanta_plugins/terraform/__init__.py", line 132, in get_resource_attribute
    raise PluginException(str(e))
inmanta.plugins.PluginException: A string attribute can only get an item from a dict, but value is of type <class 'NoneType'>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/app.py", line 799, in app
    options.func(options)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/app.py", line 577, in export
    raise exp
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/app.py", line 562, in export
    (types, scopes) = do_compile()
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 76, in do_compile
    compiler.handle_exception(e)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 248, in handle_exception
    self._handle_exception_export(e)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 254, in _handle_exception_export
    raise exception
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 246, in handle_exception
    self._handle_exception_datatrace(exception)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 258, in _handle_exception_datatrace
    raise exception
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/compiler/__init__.py", line 72, in do_compile
    success = sched.run(compiler, statements, blocks)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/execute/scheduler.py", line 410, in run
    next.execute()
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/ast/statements/call.py", line 289, in execute
    raise e
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/ast/statements/call.py", line 285, in execute
    self.function.resume(requires, self.resolver, self.queue, self.result)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/ast/statements/call.py", line 128, in resume
    self.function.call_in_context(arguments, kwargs, resolver, queue, result)
  File "/opt/inmanta/lib64/python3.9/site-packages/inmanta/ast/statements/call.py", line 259, in call_in_context
    raise ExplicitPluginException(self.ast_node, "PluginException in plugin %s" % self.ast_node.name, e)
inmanta.ast.ExplicitPluginException: PluginException in plugin terraform::get_resource_attribute
A string attribute can only get an item from a dict, but value is of type <class 'NoneType'> (reported in terraform::get_resource_attribute(self.terraform_resource,List()) (/var/lib/inmanta/docker/model/_init.cf:167))

```

</details>

Because we never expect to get this kind of value from the facts, and there is no way to recover from that.

Even though 1. is the root issue and 2. should never happen, I think we should do the following:
- Check when receiving the new state from create/update that it is not null before setting it.
- Make sure that if a state is null, the plugins don't fail on it, but simply threat the value it was trying to get as unknown. And log a warning, as this should not happen.

### PR DESCRIPTION
**Pull request by edvgui**
# Description

Deal with null states.  (more explanations in the ticket)
Closes #47

### CODE REVIEW

File: plugins/__init__.py [hunk 1]
```diff
@@ -124,6 +127,20 @@ def get_resource_attribute(
         return Unknown(source=resource)
 
     resource_state = json.loads(resource_state_raw)
+    if resource_state is None:
```
Comment (edvgui) [inline plugins/__init__.py:130]: I am having second thoughts about dealing with the null state which is never supposed to happen here, isn't it a bit redundant?

Comment (wouterdb) [reply plugins/__init__.py:130]: I'm a bit confused as well. If it should not happen, it should be a hard failure if it does. 

The question is: did it happen? If so, is the assumption that it should not happen valid.

Comment (edvgui) [reply plugins/__init__.py:130]: It happened before I fixed the behavior in the handler, not after that.  So I will leave it be a hard failure.

File: plugins/tf/terraform_resource_client.py [hunk 2]
```diff
@@ -122,14 +131,25 @@ def import_resource(self, id: str) -> Optional[dict]:
...
-            msgpack.unpackb(imported[0].state.msgpack)
-        )
         self.resource_state.private = imported[0].private
 
+        # Sanity check, the new state here should never be none
+        new_state = parse_response(msgpack.unpackb(imported[0].state.msgpack))
+        if new_state is not None:
+            self.resource_state.state = new_state
+        else:
+            raise PluginResponseException(
```
[the remark this answers is not in the document]
Comment (edvgui) [reply plugins/tf/terraform_resource_client.py:143]: So this should never happen, but as I have read in a very interesting article lately, we should always assert our assumptions (:wink: ) (so I would rather have something to fail explicitly here, than later in a much cryptic way)

File: plugins/tf/terraform_resource_client.py [hunk 3]
```diff
@@ -149,18 +169,31 @@ def read_resource(self) -> Optional[dict]:
...
-            msgpack.unpackb(result.new_state.msgpack)
-        )
         self.resource_state.private = result.private
 
+        # Sanity check, the new state here should never be none
+        new_state = parse_response(msgpack.unpackb(result.new_state.msgpack))
+        if new_state is not None:
+            self.resource_state.state = new_state
+        else:
+            raise PluginResponseException(
```
[the remark this answers is not in the document]
Comment (edvgui) [reply plugins/tf/terraform_resource_client.py:184]: Yes, which is why we raise an explicit exception about it

File: plugins/tf/terraform_resource_client.py [hunk 4]
```diff
@@ -266,16 +311,31 @@ def update_resource(self, desired: dict) -> Optional[dict]:
...
+        new_state = parse_response(msgpack.unpackb(result.new_state.msgpack))
+        if new_state is not None:
+            self.resource_state.state = new_state
+
         raise_for_diagnostics(result.diagnostics, "Failed to update the resource")
 
+        if new_state is None:
+            raise PluginResponseException(
+                "Invalid response from provider for ApplyResourceChange when updating resource.  "
+                "Received null state, this can not happen."
```
[the remark this answers is not in the document]
Comment (edvgui) [reply plugins/tf/terraform_resource_client.py]: I changed it to MUST NOT (from https://datatracker.ietf.org/doc/html/rfc2119), I think it fits?

**Review by wouterdb — APPROVED** [review]

**Pull request merged**